### ISSUE TITLE
digitalbazaar/http-client

### ISSUE THREAD

**Issue by malins** [digitalbazaar/jsonld.js#516]
Hello,

I'm having problems to test my code using Jest because jsonld library is using digitalbazaar/http-client, which in turn uses ky-universal/ky, which is only available as ESM build and does a dynamic import.

The underlying V8/node bug is described here: https://github.com/nodejs/node/issues/35889

What's the reason to use digitalbazaar/http-client and not plain fetch (with node-fetch on nodejs)? digitalbazaar/http-client seems to be quite a exotic dependency of your package.

Thank you!

**Comment by davidlehn** [comment]
As I [suggested elsewhere](https://github.com/digitalbazaar/jsonld.js/issues/451#issuecomment-1231062141), you might try the solution they recommend: https://jestjs.io/docs/ecmascript-modules.

We use our own `http-client` library in many of our other projects and it fits our use case and historically handled some compatibility issues like using `node-fetch`.  A `fetch` based `documentLoader` is certainly possible and you can write and use one with this library.  We haven't spent the effort on making one ourselves.  There might be an issue of it always loading the default, and hitting the issue you see, so maybe some overrides would be needed to stop that.

I think some `documentLoader` code refactoring is needed in general, to address use cases like this and other issues, but it hasn't been a priority.

**Comment by Peeja** [comment]
I'm having the same issue, and `m-ld` in general has run into this. `m-ld` currently has [its own fork](https://github.com/m-ld/jsonld.js) of `jsonld.js` to solve this, but it's obviously not super well maintained. The only thing it does is leave out the `http-client`.

The best solution would probably be for `documentLoader` to support `fetch`, but barring that, an intermediate solution would be for `http-client` to be optional and only used if you attempt to use the default `documentLoader`.

**Comment by dlongley** [comment]
@Peeja,

I agree that it would be a good idea to reduce the number of libraries required here (or do some kind of lazy loading) -- provided that it doesn't significantly increase the maintenance burden. That's usually a good practice in general.

However, I believe that this extra effort also shouldn't be a pressing matter here or elsewhere in the JavaScript ecosystem; elevating its priority diverts efforts that are being made in other important areas. I certainly sympathize with struggling developers -- but the cause of this perceived need for prioritization should be examined to see if there's a more efficient solution.

In my view, the best solution is for any modern tooling that purports to work with JavaScript but does not support features of the ECMAScript standard (such as ESM) should be updated such that they do. It's certainly less efficient for developers to go around either forking and maintaining or requesting modifications to every ESM library they use because it won't run with their tooling -- especially when those libraries follow the ECMAScript standard.

It seems to me that whatever tooling (usually TypeScript and / or Jest) is suffering from a lack of ESM support should be updated post-haste -- and the rest of the decentralized ecosystem that is otherwise compliant could be left alone, as it would "just work".

**Comment by dlongley** [comment]
@Peeja,

> It's not TypeScript or Jest, though, it's Node.

And Node community would say it's v8 (and they wouldn't be wrong). My view is that each of those places is a potential bottleneck for addressing the problem. Clearly the most ideal fix is in v8 -- it solves the problem for everyone (once brought into Node, etc.). The next best place for a fix would be Node, and then Jest after that. The worst place (bottleneck-wise) to do the fix would be every ESM library that any developer using Jest depends on (directly or indirectly).

> I'm not sure it's reasonable to expect Node itself to move any faster than it already is in that direction.

Sure, I'm with you on that. But, given that this problem has now been an issue for at least 2 years, I think the Jest community should move to provide some fix / temporary alternative mode to address the issue. Other testing frameworks (and many applications) are able to work with ESM without issue, including running code in Node. I think that's better than struggling developers going from library to library asking each maintainer to make changes to accommodate Jest (+/- Node +/- v8). Hopefully my position makes sense. Clearly none of this ideal.

**Comment by Peeja** [comment]
@dlongley I hear what you're saying. I'm not clear what you think the pragmatic, short term solution is, though? Just…don't just `jsonld.js` at all for a year or two? Keep maintaining our own fork? I'm not sure how else to get around it.

As an entirely separate matter, though, it's frustrating that there's no way to use `jsonld.js` without bringing along an HTTP client.

**Comment by Peeja** [comment]
Oh, hang on, I think I see the problem! This `import()` isn't in an ESM module at all: it's in [a _CommonJS_ file](https://unpkg.com/@digitalbazaar/http-client@3.3.0/dist/cjs/index.cjs). I think that's throwing off downstream tools which don't expect to do ESM-y things with that module, even if they support them. Is there maybe something wrong with the build chain that's producing that?

**Comment by dlongley** [comment]
@Peeja,

Thanks, great analysis!

It seems that a PR to http-client could be made that did something to the internals ... like making  `_createHttpClient` take a function that returned a (possibly memoized) `kyPromise` rather than taking a promise directly. That would allow the original import of `ky-universal` to also be moved within that function for the default case where the `parent` is the original / base `ky` instance:

https://github.com/digitalbazaar/http-client/blob/v3.3.0/lib/httpClient.js#L35-L51

Unfortunately, it seems this would mean that the `kyPromise` exported by the library would also need to become such a function, which would be a breaking change. Maybe we can figure out a way around that -- but even if we can't, a major version could be released that avoids the problem you identified.

Any further discussion that relates to http-client specifically should be taken over here: https://github.com/digitalbazaar/http-client/issues/34

**Comment by davidlehn** [comment]
The deferred dynamic import fix is in `@digitalbazaar/http-client@3.4.0`.    It would be helpful for people to test and see if it helps address Jest related issues.

**Comment by dmitrizagidulin** [comment]
> The deferred dynamic import fix is in `@digitalbazaar/http-client@3.4.0`. It would be helpful for people to test and see if it helps address Jest related issues.

I'm testing > `@digitalbazaar/http-client@3.4.0` (4.x actually) in Jest, and it's causing a segmentation fault due to the dynamic import :( (Still giving me a `A dynamic import callback was invoked without --experimental-vm-modules` error, even though I run jest via `NODE_OPTIONS=--experimental-modules`.
